server/fleet/software_installer.go

```diff
@@ -139,3 +139,22 @@ type HostSoftwareInstalledVersion struct {
 	Vulnerabilities []string   `json:"vulnerabilities" db:"vulnerabilities"`
 	InstalledPaths  []string   `json:"installed_paths" db:"installed_paths"`
 }
+
+// HostSoftwareInstallResultPayload is the payload provided by fleetd to record
+// the results of a software installation attempt.
+type HostSoftwareInstallResultPayload struct {
+	HostID      uint   `json:"host_id"`
+	InstallUUID string `json:"install_uuid"`
+
+	// the following fields are nil-able because the corresponding steps may not
+	// have been executed (optional step, or executed conditionally to a previous
+	// step).
+	//
+	// Note that Go's json Marshal/Unmarshal automatically convert to/from base64
+	// when the data type is []byte (https://go.dev/play/p/9cdt6impziw).
+	PreInstallConditionOutput *string `json:"pre_install_condition_output"`
+	InstallScriptExitCode     *int    `json:"install_script_exit_code"`
+	InstallScriptOutput       []byte  `json:"install_script_output"`
+	PostInstallScriptExitCode *int    `json:"post_install_script_exit_code"`
+	PostInstallScriptOutput   []byte  `json:"post_install_script_output"`
```

Comment:
Heads-up, important to note here that the API expects base64-encoded output for install script and post-install script! That's per the spec (https://github.com/fleetdm/fleet/issues/18675) but important to remember for the caller part. Also note that pre-install condition output does _not_ expect base64 (presumably because we don't expect any binary data in this output, though there shouldn't be in any output?).

Feel free to update to plain string if it was spec'd as base64 by mistake.

Reply:
good point! do you think we should do everything base64 or everything string?

Reply:
I was thinking everything string as I'm not sure there's a benefit here for base64, unless we're worried that script output may contain binary? I doubt that would happen or if it does, that it would be super important to store byte-perfect (although my understanding is that they should be properly escaped in a json string so it would still work)? WOuld be worth testing if we think that's a possibility.

Reply:
that makes sense to me, I can update that tomorrow for you. cc: @dantecatalfamo we should probably do the same for script contents.